File: libs/statemanager/src/main/kotlin/mono/state/MainStateManager.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.state

import mono.actionmanager.ActionManager
import mono.actionmanager.OneTimeActionType
import mono.actionmanager.RetainableActionType
import mono.bitmap.manager.MonoBitmapManager
import mono.common.MouseCursor
import mono.common.currentTimeMillis
import mono.common.post
import mono.environment.Build
import mono.graphics.board.Highlight
import mono.graphics.board.MonoBoard
import mono.graphics.geo.DirectedPoint
import mono.graphics.geo.MousePointer
import mono.graphics.geo.Point
import mono.graphics.geo.Rect
import mono.html.canvas.CanvasViewController
import mono.lifecycle.LifecycleOwner
import mono.livedata.LiveData
import mono.livedata.MutableLiveData
import mono.livedata.distinctUntilChange
import mono.shape.ShapeManager
import mono.shape.add
import mono.shape.clipboard.ShapeClipboardManager
import mono.shape.connector.ShapeConnector
import mono.shape.remove
import mono.shape.selection.SelectedShapeManager
import mono.shape.selection.SelectedShapeManager.ShapeFocusType
import mono.shape.shape.AbstractShape
import mono.shape.shape.Group
import mono.shape.shape.Line
import mono.shape.shape.MockShape
import mono.shape.shape.Rectangle
import mono.shape.shape.RootGroup
import mono.shape.shape.Text
import mono.shapebound.InteractionPoint
import mono.shapebound.LineInteractionBound
import mono.shapebound.ScalableInteractionBound
import mono.shapesearcher.ShapeSearcher
import mono.state.command.CommandEnvironment
import mono.state.command.CommandEnvironment.EditingMode
import mono.state.command.MouseCommandFactory
import mono.state.command.mouse.MouseCommand
import mono.store.dao.workspace.WorkspaceDao
import mono.ui.appstate.AppUiStateManager

/**
 * A class which connects components in the app.
 */
class MainStateManager(
    lifecycleOwner: LifecycleOwner,
    private val mainBoard: MonoBoard,
    private val shapeManager: ShapeManager,
    private val selectedShapeManager: SelectedShapeManager,
    private val bitmapManager: MonoBitmapManager,
    private val canvasManager: CanvasViewController,
    shapeClipboardManager: ShapeClipboardManager,
    mousePointerLiveData: LiveData<MousePointer>,
    applicationActiveStateLiveData: LiveData<Boolean>,
    private val actionManager: ActionManager,
    appUiStateManager: AppUiStateManager,
    initialRootId: String = "",
    private val workspaceDao: WorkspaceDao = WorkspaceDao.instance
) {
    private val shapeSearcher: ShapeSearcher = ShapeSearcher(shapeManager, bitmapManager::getBitmap)

    /**
     * The current working parent group, which is the group that is focused, shape actions will be
     * applied to this group.
     *
     * This is similar to the concept of "current directory" in file system.
     */
    private var workingParentGroup: Group = shapeManager.root

    private var windowBoardBound: Rect = Rect.ZERO

    private val environment = CommandEnvironmentImpl(this)

    private var currentMouseCommand: MouseCommand? = null
    private var currentRetainableActionType: RetainableActionType = RetainableActionType.IDLE

    private val redrawRequestMutableLiveData = MutableLiveData(Unit)

    private val editingModeLiveData = MutableLiveData(EditingMode.idle(null))
    private val stateHistoryManager =
        StateHistoryManager(lifecycleOwner, environment, canvasManager)

    init {
        mousePointerLiveData
            .distinctUntilChange()
            .observe(lifecycleOwner, listener = ::onMouseEvent)

        mousePointerLiveData
            .distinctUntilChange()
            .observe(lifecycleOwner, listener = ::updateMouseCursor)

        canvasManager.windowBoardBoundLiveData.observe(lifecycleOwner) {
            windowBoardBound = it
            if (Build.DEBUG) {
                println(
                    "¶ Drawing info: window board size $windowBoardBound • " +
                        "pixel size ${canvasManager.windowBoundPx}"
                )
            }
            requestRedraw()
        }

        shapeManager.versionLiveData
            .distinctUntilChange()
            .observe(lifecycleOwner) {
                requestRedraw()
            }

        environment.selectedShapesLiveData.observe(
            lifecycleOwner,
            listener = ::updateInteractionBounds
        )

        redrawRequestMutableLiveData.observe(
            lifecycleOwner,
            throttleDurationMillis = 0
        ) { redraw() }

        actionManager.retainableActionLiveData.observe(lifecycleOwner) {
            currentRetainableActionType = it
        }

        stateHistoryManager.restoreAndStartObserveStateChange(initialRootId)

        OneTimeActionHandler(
            lifecycleOwner,
            actionManager.oneTimeActionLiveData,
            environment,
            bitmapManager,
            shapeClipboardManager,
            stateHistoryManager,
            appUiStateManager
        )

        applicationActiveStateLiveData.observe(lifecycleOwner) { isActive ->
            if (isActive) {
                reflectChangedFromLocal()
            }
        }
    }

    private fun onMouseEvent(mousePointer: MousePointer) {
        if (mousePointer is MousePointer.DoubleClick) {
            val targetedShape =
                environment.getSelectedShapes()
                    .firstOrNull { it.contains(mousePointer.boardCoordinate) }
            actionManager.setOneTimeAction(OneTimeActionType.EditSelectedShape(targetedShape))
            return
        }

        val mouseCommand =
            MouseCommandFactory.getCommand(environment, mousePointer, currentRetainableActionType)
                ?: currentMouseCommand
                ?: return
        currentMouseCommand = mouseCommand

        environment.enterEditingMode()
        val commandResultType = mouseCommand.execute(environment, mousePointer)

        if (commandResultType == MouseCommand.CommandResultType.DONE) {
            environment.exitEditingMode(true)
        }

        if (commandResultType == MouseCommand.CommandResultType.DONE ||
            commandResultType == MouseCommand.CommandResultType.WORKING_PHASE2
        ) {
            currentMouseCommand = null
            requestRedraw()
            // Avoid click when adding shape cause shape selection command
            post { actionManager.setRetainableAction(RetainableActionType.IDLE) }
        }
    }

    /**
     * Redraws all content on the workspace.
     * This is used when the theme is updated.
     */
    fun forceFullyRedrawWorkspace() {
        canvasManager.fullyRedraw()
    }

    private fun requestRedraw() {
        redrawRequestMutableLiveData.value = Unit
    }

    private fun redraw() {
        auditPerformance("Redraw") {
            redrawMainBoard()
        }
        auditPerformance("Draw canvas") {
            canvasManager.drawBoard()
        }
    }

    private fun redrawMainBoard() {
        shapeSearcher.clear(windowBoardBound)
        mainBoard.clearAndSetWindow(windowBoardBound)
        drawShapeToMainBoard(shapeManager.root)
    }

    private fun drawShapeToMainBoard(shape: AbstractShape) {
        if (shape is Group) {
            for (child in shape.items) {
                drawShapeToMainBoard(child)
            }
            return
        }
        val bitmap = bitmapManager.getBitmap(shape) ?: return
        val highlight = when {
            shape is Text && shape.isTextEditing -> Highlight.TEXT_EDITING
            shape in environment.getSelectedShapes() -> Highlight.SELECTED
            else -> {
                when (selectedShapeManager.getFocusingType(shape)) {
                    ShapeFocusType.LINE_CONNECTING -> Highlight.LINE_CONNECT_FOCUSING
                    null -> Highlight.NO
                }
            }
        }
        mainBoard.fill(shape.bound.position, bitmap, highlight)
        shapeSearcher.register(shape)
    }

    private fun auditPerformance(
        objective: String,
        isEnabled: Boolean = DEBUG_PERFORMANCE_AUDIT_ENABLED,
        action: () -> Unit
    ) {
        if (!isEnabled || !Build.DEBUG) {
            action()
            return
        }
        val t0 = currentTimeMillis()
        action()
        println("$objective runtime: ${currentTimeMillis() - t0}")
    }

    private fun updateMouseCursor(mousePointer: MousePointer) {
        val mouseCursor = when (mousePointer) {
            is MousePointer.Move -> getMouseMovingCursor(mousePointer)

            is MousePointer.Drag -> {
                val mouseCommand = currentMouseCommand
                if (mouseCommand != null) mouseCommand.mouseCursor else MouseCursor.DEFAULT
            }

            is MousePointer.Up -> MouseCursor.DEFAULT

            MousePointer.Idle,
            is MousePointer.Down,
            is MousePointer.Click,
            is MousePointer.DoubleClick -> null
        }
        if (mouseCursor != null) {
            canvasManager.setMouseCursor(mouseCursor)
        }
    }

    private fun getMouseMovingCursor(mousePointer: MousePointer.Move): MouseCursor {
        val interactionPoint = canvasManager.getInteractionPoint(mousePointer.pointPx)
        return interactionPoint?.mouseCursor ?: currentRetainableActionType.mouseCursor
    }

    private fun updateInteractionBounds(selectedShapes: Collection<AbstractShape>) {
        val bounds = selectedShapes.mapNotNull {
            when (it) {
                is Rectangle,
                is Text -> ScalableInteractionBound(it.id, it.bound)

                is Line -> LineInteractionBound(it.id, it.edges)

                is Group -> null // TODO: Add new Interaction bound type for Group
                is MockShape -> null
            }
        }
        canvasManager.drawInteractionBounds(bounds)
        requestRedraw()
    }

    private fun reflectChangedFromLocal() {
        val rootId = shapeManager.root.id
        val rootVersion = shapeManager.root.versionCode
        val currentRoot = workspaceDao.getObject(rootId).rootGroup
        // TODO: load from storage
        val shapeConnector = ShapeConnector()
        when {
            currentRoot == null -> {
                // TODO: Notify to user this project is removed
            }

            rootVersion != currentRoot.versionCode -> {
                environment.replaceRoot(RootGroup(currentRoot), shapeConnector)
            }
        }
    }

    private class CommandEnvironmentImpl(
        private val stateManager: MainStateManager
    ) : CommandEnvironment {
        override val shapeManager: ShapeManager
            get() = stateManager.shapeManager

        private val shapeSearcher: ShapeSearcher
            get() = stateManager.shapeSearcher

        override val editingModeLiveData: LiveData<EditingMode>
            get() = stateManager.editingModeLiveData

        override var workingParentGroup: Group
            get() = stateManager.workingParentGroup
            private set(value) {
                stateManager.workingParentGroup = value
            }

        override fun replaceRoot(newRoot: RootGroup, newShapeConnector: ShapeConnector) {
            val currentRoot = shapeManager.root
            if (currentRoot.id != newRoot.id) {
                stateManager.workspaceDao.getObject(objectId = newRoot.id).updateLastOpened()
                stateManager.canvasManager.setOffset(
                    stateManager.workspaceDao.getObject(newRoot.id).offset
                )
                stateManager.stateHistoryManager.clear()
            }

            shapeManager.replaceRoot(newRoot, newShapeConnector)
            workingParentGroup = shapeManager.root
            clearSelectedShapes()
        }

        override fun enterEditingMode() {
            stateManager.editingModeLiveData.value = EditingMode.edit()
        }

        override fun exitEditingMode(isNewStateAccepted: Boolean) {
            val skippedVersion =
                if (isNewStateAccepted) null else shapeManager.versionLiveData.value
            stateManager.editingModeLiveData.value = EditingMode.idle(skippedVersion)
        }

        override fun addShape(shape: AbstractShape?) {
            if (shape != null) {
                shapeManager.add(shape)
            }
        }

        override fun removeShape(shape: AbstractShape?) = shapeManager.remove(shape)

        override fun getShapes(point: Point): Sequence<AbstractShape> =
            shapeSearcher.getShapes(point)

        override fun getAllShapesInZone(bound: Rect): Sequence<AbstractShape> =
            shapeSearcher.getAllShapesInZone(bound)

        override fun getWindowBound(): Rect = stateManager.windowBoardBound

        override fun getInteractionPoint(pointPx: Point): InteractionPoint? =
            stateManager.canvasManager.getInteractionPoint(pointPx)

        override fun updateInteractionBounds() =
            stateManager.updateInteractionBounds(stateManager.selectedShapeManager.selectedShapes)

        override fun isPointInInteractionBounds(point: Point): Boolean =
            stateManager.selectedShapeManager.selectedShapes.any { it.contains(point) }

        override fun setSelectionBound(bound: Rect?) =
            stateManager.canvasManager.drawSelectionBound(bound)

        override val selectedShapesLiveData: LiveData<Set<AbstractShape>> =
            stateManager.selectedShapeManager.selectedShapesLiveData

        override fun getSelectedShapes(): Set<AbstractShape> =
            stateManager.selectedShapeManager.selectedShapes

        override fun addSelectedShape(shape: AbstractShape?) {
            if (shape != null) {
                stateManager.selectedShapeManager.addSelectedShape(shape)
            }
        }

        override fun toggleShapeSelection(shape: AbstractShape) =
            stateManager.selectedShapeManager.toggleSelection(shape)

        override fun setFocusingShape(
            shape: AbstractShape?,
            focusType: ShapeFocusType
        ) = stateManager.selectedShapeManager.setFocusingShape(shape, focusType)

        override fun selectAllShapes() {
            for (shape in stateManager.workingParentGroup.items) {
                addSelectedShape(shape)
            }
        }

        override fun clearSelectedShapes() = stateManager.selectedShapeManager.clearSelectedShapes()

        override fun getEdgeDirection(point: Point): DirectedPoint.Direction? =
            shapeSearcher.getEdgeDirection(point)

        override fun toXPx(column: Double): Double = stateManager.canvasManager.toXPx(column)

        override fun toYPx(row: Double): Double = stateManager.canvasManager.toYPx(row)

        override fun toWidthPx(width: Double): Double = stateManager.canvasManager.toWidthPx(width)

        override fun toHeightPx(height: Double): Double =
            stateManager.canvasManager.toHeightPx(height)
    }

    companion object {
        private const val DEBUG_PERFORMANCE_AUDIT_ENABLED = false
    }
}

File: monosketch-svelte/src/lib/mono/state-manager/main-state-manager.ts
/*
 * Copyright (c) 2024, tuanchauict
 */

import type { Workspace } from "$app/workspace";
import { Flow, LifecycleOwner } from "$libs/flow";
import type { Direction, Point } from "$libs/graphics-geo/point";
import { Rect } from "$libs/graphics-geo/rect";
import { any } from "$libs/sequence";
import { unit, type Unit } from "$libs/unit";
import type { ActionManager } from "$mono/action-manager/action-manager";
import { RetainableActionTypeMouseCursor } from "$mono/action-manager/retainable-actions";
import { DEBUG_MODE } from "$mono/build_environment";
import { MonoBoard } from "$mono/monobitmap/board";
import { HighlightType } from "$mono/monobitmap/board/pixel";
import type { MonoBitmapManager } from "$mono/monobitmap/manager/mono-bitmap-manager";
import type { InteractionPoint } from "$mono/shape-interaction-bound/interaction-point";
import { AddShape, RemoveShape } from "$mono/shape/command/shape-manager-commands";
import type { ShapeConnector } from "$mono/shape/connector/shape-connector";
import { ShapeSearcher } from "$mono/shape/searcher/shape-searcher";
import { type FocusingShape, type SelectedShapeManager, ShapeFocusType } from "$mono/shape/selected-shape-manager";
import type { ShapeClipboardManager } from "$mono/shape/shape-clipboard-manager";
import { type Command, ShapeManager } from "$mono/shape/shape-manager";
import type { AbstractShape } from "$mono/shape/shape/abstract-shape";
import { Group } from "$mono/shape/shape/group";
import { Text } from "$mono/shape/shape/text";
import { type CommandEnvironment, EditingMode } from "$mono/state-manager/command-environment";
import { MouseInteractionController } from "$mono/state-manager/controller/mouse-interaction-controller";
import { OneTimeActionHandler } from "$mono/state-manager/one-time-action-handler";
import { StateHistoryManager } from "$mono/state-manager/state-history-manager";
import type { WorkspaceDao } from "$mono/store-manager/dao/workspace-dao";
import type { AppUiStateManager } from "$mono/ui-state-manager/app-ui-state-manager";
import { MouseCursor } from "$mono/workspace/mouse/cursor-type";
import { MousePointer, MousePointerType } from "$mono/workspace/mouse/mouse-pointer";

/**
 * A class which connects components in the app.
 */
export class MainStateManager {
    private readonly commandEnvironment: CommandEnvironment;

    private workingParentGroup: Group;

    private readonly shapeSearcher: ShapeSearcher;
    private readonly stateHistoryManager: StateHistoryManager;

    private readonly oneTimeActionHandler: OneTimeActionHandler;

    private readonly redrawRequestMutableFlow: Flow<Unit> = new Flow(unit);
    private readonly mouseInteractionController: MouseInteractionController;

    constructor(
        private readonly mainBoard: MonoBoard,
        private readonly shapeManager: ShapeManager,
        private readonly selectedShapeManager: SelectedShapeManager,
        private readonly bitmapManager: MonoBitmapManager,
        private readonly workspace: Workspace,
        private readonly workspaceDao: WorkspaceDao,
        private readonly actionManager: ActionManager,
        shapeClipboardManager: ShapeClipboardManager,
        appUiStateManager: AppUiStateManager,
        initialRootId: string,
    ) {
        if (DEBUG_MODE) {
            console.log(`Root ID: ${initialRootId}`);
        }

        this.shapeSearcher = new ShapeSearcher(shapeManager, shape => bitmapManager.getBitmap(shape));
        this.commandEnvironment = new CommandEnvironmentImpl({
            mainStateManager: this,
            shapeManager: this.shapeManager,
            selectedShapeManager: this.selectedShapeManager,
            shapeSearcher: this.shapeSearcher,
            workspace: this.workspace,
            workspaceDao: this.workspaceDao,

            getWorkingParentGroup: () => this.workingParentGroup,
            setWorkingParentGroup: (group: Group) => this.workingParentGroup = group,

            replaceRoot: (newRoot: Group, newShapeConnector: ShapeConnector) => {
                this.replaceRoot(newRoot, newShapeConnector);
            },
        });

        this.stateHistoryManager = new StateHistoryManager(this.commandEnvironment, this.workspace, this.workspaceDao);
        this.stateHistoryManager.restoreState(initialRootId);

        this.workingParentGroup = shapeManager.root;

        this.oneTimeActionHandler = new OneTimeActionHandler(
            this.commandEnvironment,
            this.bitmapManager,
            shapeClipboardManager,
            this.stateHistoryManager,
            appUiStateManager,
            this.workspaceDao,
        );

        this.mouseInteractionController = new MouseInteractionController(
            this.commandEnvironment,
            this.actionManager,
            () => this.requestRedraw(),
        );
    }

    get windowBoardBound(): Rect {
        return this.workspace.windowBoardBoundFlow.value ?? Rect.ZERO;
    }

    onStart(lifecycleOwner: LifecycleOwner): void {
        this.workspace.drawingOffsetPointPxFlow.observe(lifecycleOwner, (offsetPointPx) => {
            this.workspaceDao.getObject(this.shapeManager.root.id).offset = offsetPointPx;
        });
        this.workspace.windowBoardBoundFlow.observe(lifecycleOwner, (windowBoardBound) => {
            if (DEBUG_MODE) {
                console.log(`¶ Drawing info: window board size ${windowBoardBound} • pixel size ${this.workspace.getDrawingInfo().boundPx}`);
            }
            this.requestRedraw();
        });

        this.shapeManager.rootIdFlow.observe(lifecycleOwner, (rootId) => {
            this.workspace.setDrawingOffset(this.workspaceDao.getObject(rootId).offset);
        });
        this.shapeManager.versionFlow.observe(lifecycleOwner, (version) => {
            this.requestRedraw();
        });

        this.stateHistoryManager.observeStateChange(lifecycleOwner);
        this.oneTimeActionHandler.observe(lifecycleOwner, this.actionManager.oneTimeActionFlow);

        this.redrawRequestMutableFlow.throttle(0).observe(lifecycleOwner, () => this.redraw());

        this.workspace.mousePointerFlow.distinctUntilChanged().observe(lifecycleOwner, (mousePointer) => {
            this.mouseInteractionController.onMouseEvent(mousePointer);
            this.updateMouseCursor(mousePointer);
        });
    }

    replaceRoot(newRoot: Group, newShapeConnector: ShapeConnector): void {
        const currentRoot = this.shapeManager.root;
        if (currentRoot.id !== newRoot.id) {
            this.workspaceDao.getObject(newRoot.id).lastModifiedTimestampMillis = Date.now();
            this.workspace.setDrawingOffset(
                this.workspaceDao.getObject(newRoot.id).offset,
            );
            this.stateHistoryManager.clear();
        }

        this.shapeManager.replaceRoot(newRoot, newShapeConnector);
        this.workingParentGroup = this.shapeManager.root;
        this.commandEnvironment.clearSelectedShapes();
    }

    private requestRedraw() {
        this.redrawRequestMutableFlow.value = unit;
    }

    private redraw() {
        this.redrawMainBoard();
        this.workspace.draw();
    }

    private redrawMainBoard() {
        const windowBoardBound = this.windowBoardBound;
        this.shapeSearcher.clear(windowBoardBound);
        this.mainBoard.clearAndSetWindow(windowBoardBound);
        this.drawShapeToMainBoard(this.shapeManager.root);
    }

    private drawShapeToMainBoard(shape: AbstractShape) {
        if (shape instanceof Group) {
            for (const child of shape.items) {
                this.drawShapeToMainBoard(child);
            }
            return;
        }

        const bitmap = this.bitmapManager.getBitmap(shape);
        if (!bitmap) {
            return;
        }
        this.mainBoard.fillBitmap(shape.bound.position, bitmap, this.getHighlightType(shape));
        this.shapeSearcher.register(shape);
    }

    private getHighlightType(shape: AbstractShape): HighlightType {
        if (shape instanceof Text && shape.isTextEditing) {
            return HighlightType.TEXT_EDITING;
        }
        if (this.commandEnvironment.getSelectedShapes().has(shape)) {
            return HighlightType.SELECTED;
        }
        const focusingType = this.selectedShapeManager.getFocusingType(shape);
        switch (focusingType) {
            case ShapeFocusType.LINE_CONNECTING:
                return HighlightType.LINE_CONNECT_FOCUSING;
            case ShapeFocusType.SELECT_MODE_HOVER:
                return HighlightType.SELECTED;
            case null:
                return HighlightType.NO;
        }
    }

    private updateMouseCursor(mousePointer: MousePointer) {
        const mouseCursor = (() => {
            switch (mousePointer.type) {
                case MousePointerType.MOVE: {
                    const interactionPoint = this.commandEnvironment.getInteractionPoint(mousePointer.clientCoordinate);
                    return interactionPoint?.mouseCursor
                        ?? RetainableActionTypeMouseCursor[this.mouseInteractionController.currentRetainableActionType];
                }
                case MousePointerType.DRAG: {
                    const mouseCommand = this.mouseInteractionController.activeMouseCommand;
                    return mouseCommand?.mouseCursor ?? MouseCursor.DEFAULT;
                }
                case MousePointerType.UP:
                    return MouseCursor.DEFAULT;
                default:
                    return null;
            }
        })();
        if (mouseCursor) {
            this.workspace.setMouseCursor(mouseCursor);
        }
    }

    updateInteractionBounds(selectedShapes: Set<AbstractShape>): void {
        throw new Error("Method not implemented.");
    }
}

interface DependencyManager {
    readonly mainStateManager: MainStateManager;
    readonly shapeManager: ShapeManager;
    readonly selectedShapeManager: SelectedShapeManager;
    readonly shapeSearcher: ShapeSearcher;
    readonly workspace: Workspace;
    readonly workspaceDao: WorkspaceDao;

    getWorkingParentGroup(): Group;

    setWorkingParentGroup(group: Group): void;

    replaceRoot(newRoot: Group, newShapeConnector: ShapeConnector): void;
}

class CommandEnvironmentImpl implements CommandEnvironment {
    shapeManager: ShapeManager;

    editingModeFlow: Flow<EditingMode> = new Flow(EditingMode.idle(null));

    constructor(
        private readonly dependencies: DependencyManager,
    ) {
        this.shapeManager = dependencies.shapeManager;
    }

    get workingParentGroup(): Group {
        return this.dependencies.getWorkingParentGroup();
    };

    set workingParentGroup(group: Group) {
        this.dependencies.setWorkingParentGroup(group);
    }

    get currentRootGroup(): Group {
        return this.shapeManager.root;
    };

    replaceRoot(newRoot: Group, newShapeConnector: ShapeConnector): void {
        this.dependencies.replaceRoot(newRoot, newShapeConnector);
    }

    enterEditingMode(): void {
        this.editingModeFlow.value = EditingMode.edit();
    }

    exitEditingMode(isNewStateAccepted: boolean): void {
        const skippedVersion = isNewStateAccepted ? null : this.shapeManager.versionFlow.value ?? null;
        this.editingModeFlow.value = EditingMode.idle(skippedVersion);
    }

    executeShapeManagerCommand(command: Command): void {
        this.dependencies.shapeManager.execute(command);
    }

    addShape(shape: AbstractShape | null): void {
        if (shape) {
            this.executeShapeManagerCommand(new AddShape(shape));
        }
    }

    removeShape(shape: AbstractShape | null): void {
        if (shape) {
            this.executeShapeManagerCommand(new RemoveShape(shape));
        }
    }

    getShapes(point: Point): AbstractShape[] {
        return this.dependencies.shapeSearcher.getShapes(point);
    }

    getAllShapesInZone(bound: Rect): Iterable<AbstractShape> {
        return this.dependencies.shapeSearcher.getAllShapesInZone(bound);
    }

    getWindowBound(): Rect {
        return this.dependencies.workspace.windowBoardBoundFlow.value ?? Rect.ZERO;
    }

    getInteractionPoint(pointPx: Point): InteractionPoint | null {
        return this.dependencies.workspace.getInteractionPoint(pointPx);
    }

    updateInteractionBounds(): void {
        this.dependencies.mainStateManager.updateInteractionBounds(this.dependencies.selectedShapeManager.selectedShapes);
    }

    isPointInInteractionBounds(point: Point): boolean {
        return any(this.dependencies.selectedShapeManager.selectedShapes, shape => shape.contains(point));
    }

    setSelectionBound(bound: Rect | null): void {
        this.dependencies.workspace.drawSelectionBound(bound);
    }

    get selectedShapesFlow(): Flow<Set<AbstractShape>> {
        return this.dependencies.selectedShapeManager.selectedShapesFlow;
    }

    getSelectedShapes(): Set<AbstractShape> {
        return this.dependencies.selectedShapeManager.selectedShapes;
    }

    addSelectedShape(shape: AbstractShape | null): void {
        if (shape) {
            this.dependencies.selectedShapeManager.addSelectedShape(shape);
        }
    }

    toggleShapeSelection(shape: AbstractShape): void {
        this.dependencies.selectedShapeManager.toggleSelection(shape);
    }

    setFocusingShape(shape: AbstractShape | null, focusType: ShapeFocusType): void {
        this.dependencies.selectedShapeManager.setFocusingShape(shape, focusType);
    }

    getFocusingShape(): FocusingShape | null {
        return this.dependencies.selectedShapeManager.focusingShape;
    }

    selectAllShapes(): void {
        for (const shape of this.dependencies.getWorkingParentGroup().items) {
            this.addSelectedShape(shape);
        }
    }

    clearSelectedShapes(): void {
        this.dependencies.selectedShapeManager.clearSelectedShapes();
    }

    getEdgeDirection(point: Point): Direction | null {
        return this.dependencies.shapeSearcher.getEdgeDirection(point);
    }

    toXPx(column: number): number {
        return this.dependencies.workspace.getDrawingInfo().toXPx(column);
    }

    toYPx(row: number): number {
        return this.dependencies.workspace.getDrawingInfo().toYPx(row);
    }

    toWidthPx(width: number): number {
        return this.dependencies.workspace.getDrawingInfo().toWidthPx(width);
    }

    toHeightPx(height: number): number {
        return this.dependencies.workspace.getDrawingInfo().toHeightPx(height);
    }
}
